### ISSUE TITLE
Filtering by SrcSysID does not function as intended (bug)

### ISSUE THREAD

**Issue by max-itsl** [mavlink-router/mavlink-router#368]
Hello

**TLDR: AllowSrcSysIn filter does not work at all, but AllowSrcSysOut works as a two-way whitelist filter (IN and OUT).**

**Test setup scheme:**

![mavrouter test setup for github issue](https://user-images.githubusercontent.com/66274309/181103630-635205c9-0042-400c-bc30-5ae622e0fe9d.png)

Firmware: latest PX4 1.13 stable
Hardware: Cube Black, Cube Orange, Holybro Pixhawk 4, Matek H743 Slim
GCS: Recent QGroundControl Daily build

### Test case 1: No filtering applied

Configuration file content /etc/mavlink-router/main.conf  :
```
[General]
TcpServerPort = 5760
[UartEndpoint ToRover15]
Device = /dev/ttyTHS1
Baud = 57600
```

**Test 1 result:** Everything works fine. Can easily connect from both instances of QGroundControl (SysID 233 and 241 at the same time). Parameters download procedure is very fast. Can upload and download missions.

Debug console output: sudo mavlink-routerd -g debug
```
mavlink-router version v2-229-g022333d
Opened UART [4]ToRover15: /dev/ttyTHS1
UART [4]ToRover15: speed = 57600
Opened TCP Server [5] [::]:5760
…
…
TCP Server: New client
TCP [7]dynamic: Connection accepted
…
…
> UART [4]ToRover15: Got 40 bytes
Endpoint [4]ToRover15: got message 30 to -1/-1 from 15/1
	Known components:
		15/1
Endpoint [7]dynamic: got message 30 to -1/-1 from 15/1
	Known components:
Endpoint [7] accepted message 30 to -1/-1 from 15/1
TCP [7]dynamic: Wrote 40 bytes
…
…
> TCP [7]dynamic: Got 21 bytes
Endpoint [4]ToRover15: got message 0 to -1/-1 from 233/190
	Known components:
		15/1
Endpoint [4] accepted message 0 to -1/-1 from 233/190
UART [4]ToRover15: Wrote 21 bytes
Endpoint [7]dynamic: got message 0 to -1/-1 from 233/190
	Known components:
		233/190
…
…
TCP Server: New client
TCP [8]dynamic: Connection accepted
> UART [4]ToRover15: Got 29 bytes
Endpoint [4]ToRover15: got message 74 to -1/-1 from 15/1
	Known components:
		15/1
Endpoint [7]dynamic: got message 74 to -1/-1 from 15/1
	Known components:
		233/190
Endpoint [7] accepted message 74 to -1/-1 from 15/1
TCP [7]dynamic: Wrote 29 bytes
Endpoint [8]dynamic: got message 74 to -1/-1 from 15/1
	Known components:
Endpoint [8] accepted message 74 to -1/-1 from 15/1
TCP [8]dynamic: Wrote 29 bytes
…
…
> TCP [8]dynamic: Got 65 bytes
Endpoint [4]ToRover15: got message 76 to 15/1 from 241/190
	Known components:
		15/1
Endpoint [4] accepted message 76 to 15/1 from 241/190
UART [4]ToRover15: Wrote 44 bytes
Endpoint [7]dynamic: got message 76 to 15/1 from 241/190
	Known components:
		233/190
Endpoint [8]dynamic: got message 76 to 15/1 from 241/190
	Known components:
		241/190
Endpoint [4]ToRover15: got message 0 to -1/-1 from 241/190
	Known components:
		15/1
Endpoint [4] accepted message 0 to -1/-1 from 241/190
UART [4]ToRover15: Wrote 21 bytes
Endpoint [7]dynamic: got message 0 to -1/-1 from 241/190
	Known components:
		233/190
Endpoint [7] accepted message 0 to -1/-1 from 241/190
TCP [7]dynamic: Wrote 21 bytes
Endpoint [8]dynamic: got message 0 to -1/-1 from 241/190
	Known components:
		241/190
…
…


```

### Test case 2: Applied “AllowSrcSysIn” filter to UART
**The goal was to try disabling / enabling incoming communication with Rover (SysID 15) to test “input” filter.**

Configuration file content /etc/mavlink-router/main.conf :
```
[General]
TcpServerPort = 5760
[UartEndpoint ToRover15]
Device = /dev/ttyTHS1
Baud = 57600
AllowSrcSysIn = 11 
```

_**Comment to the config file**: 
used “= 11” as arbitrary value, meaning “not 15 or 233 or 241”. 
Also tried: 
AllowSrcSysIn = 1,255
AllowSrcSysIn = 1,2,3,4,15,233,241,255
AllowSrcSysIn = 15
AllowSrcSysIn = 15,233,241_

**Test 2 result:** **Input filter seems not working.** Every attempt – the same result as during Test 1 - everything works fine. Mavlink traffic passes without restrictions. Can easily connect from both instances of QGroundControl (SysID 233 and 241 at the same time). Parameters download procedure is very fast. Can upload and download missions.

Debug console output: **the same as during Test 1.**

### Test case 3: Applied “AllowSrcSysOut” filter to UART
**The goal was to try to isolate one of GCS (SysID 241)  from Rover (SysID 15) .**

Configuration file content /etc/mavlink-router/main.conf :
```
[General]
TcpServerPort = 5760
[UartEndpoint ToRover15]
Device = /dev/ttyTHS1
Baud = 57600
AllowSrcSysOut = 233
```

**Test 3 result: wrong OUT filter behavior, acts as an incoming filter (as debug console says).** No connection to Rover from any QGC. Messages from Rover (SysID 15) are blocked – see debug console. But QGroundControl 233 sees QGroundControl 241 using Mavlink Inspector (via dynamic TCP endpoints).

Debug console output: sudo mavlink-routerd -g debug
```
mavlink-router version v2-229-g022333d
Opened UART [4]ToRover15: /dev/ttyTHS1
UART [4]ToRover15: speed = 57600
Opened TCP Server [5] [::]:5760
> UART [4]ToRover15: Got 82 bytes
Message 340 to -1/-1 **from 15/1 discarded by incoming filters**
> UART [4]ToRover15: Got 40 bytes
Message 30 to -1/-1 **from 15/1 discarded by incoming filters**
> UART [4]ToRover15: Got 29 bytes
Message 74 to -1/-1 **from 15/1 discarded by incoming filters**
> UART [4]ToRover15: Got 40 bytes
Message 33 to -1/-1 **from 15/1 discarded by incoming filters**
…
…
> UART [4]ToRover15: Got 40 bytes
Message 30 to -1/-1 from 15/1 discarded by incoming filters
> TCP [7]dynamic: Got 21 bytes
Endpoint [4]ToRover15: got message 0 to -1/-1 from 233/190
	Known components:
Endpoint [4] accepted message 0 to -1/-1 from 233/190
UART [4]ToRover15: Wrote 21 bytes
Endpoint [7]dynamic: got message 0 to -1/-1 from 233/190
	Known components:
		233/190
> UART [4]ToRover15: Got 40 bytes
Message 33 to -1/-1 from 15/1 discarded by incoming filters
…
…
> UART [4]ToRover15: Got 40 bytes
Message 30 to -1/-1 from 15/1 discarded by incoming filters
> TCP [7]dynamic: Got 21 bytes
Endpoint [4]ToRover15: got message 0 to -1/-1 from 233/190
	Known components:
Endpoint [4] accepted message 0 to -1/-1 from 233/190
UART [4]ToRover15: Wrote 21 bytes
Endpoint [7]dynamic: got message 0 to -1/-1 from 233/190
	Known components:
		233/190
Endpoint [8]dynamic: got message 0 to -1/-1 from 233/190
	Known components:
		241/190
Endpoint [8] accepted message 0 to -1/-1 from 233/190
TCP [8]dynamic: Wrote 21 bytes
> UART [4]ToRover15: Got 40 bytes
Message 30 to -1/-1 from 15/1 discarded by incoming filters
…
…


```

### Test case 4: allow SysID 15
**As OUT filter acts as an IN filter for UART port, let’s allow Rover SysID 15 as it was previously blocked. We already have strange behavior, so now we explore.**

Configuration file /etc/mavlink-router/main.conf  content:
```
[General]
TcpServerPort = 5760
[UartEndpoint ToRover15]
Device = /dev/ttyTHS1
Baud = 57600
AllowSrcSysOut = 15
```

**Test 4 result:** now both QGroundControl instances receive telemetry from Rover, but cannot retrieve any parameters and show error: _“Vehicle 15 did not respond to request for parameters. This will cause QGroundControl Daily to be unable to display its full user interface”_. Seems like **one-way communication…**

Debug console output: sudo mavlink-routerd -g debug
```
mavlink-router version v2-229-g022333d
Opened UART [4]ToRover15: /dev/ttyTHS1
UART [4]ToRover15: speed = 57600
Opened TCP Server [5] [::]:5760
> UART [4]ToRover15: Got 191 bytes
Endpoint [4]ToRover15: got message 24 to -1/-1 from 15/1
	Known components:
		15/1
Message 24 to unknown sysid/compid: -1/-1
Endpoint [4]ToRover15: got message 32 to -1/-1 from 15/1
	Known components:
		15/1
Message 32 to unknown sysid/compid: -1/-1
Endpoint [4]ToRover15: got message 1 to -1/-1 from 15/1
	Known components:
		15/1
Message 1 to unknown sysid/compid: -1/-1
Endpoint [4]ToRover15: got message 8 to -1/-1 from 15/1
	Known components:
		15/1
Message 8 to unknown sysid/compid: -1/-1
…
…
> TCP [7]dynamic: Got 17 bytes
Endpoint [4]ToRover15: got message 0 to -1/-1 from 233/190
	Known components:
		15/1
**Endpoint [4] filtered out message 0 to -1/-1 from 233/190**
Endpoint [7]dynamic: got message 0 to -1/-1 from 233/190
	Known components:
		233/190
Endpoint [8]dynamic: got message 0 to -1/-1 from 233/190
	Known components:
		241/190
Endpoint [8] accepted message 0 to -1/-1 from 233/190
TCP [8]dynamic: Wrote 17 bytes
> UART [4]ToRover15: Got 29 bytes
Endpoint [4]ToRover15: got message 74 to -1/-1 from 15/1
	Known components:
		15/1
Endpoint [7]dynamic: got message 74 to -1/-1 from 15/1
	Known components:
		233/190
Endpoint [7] accepted message 74 to -1/-1 from 15/1
TCP [7]dynamic: Wrote 29 bytes
Endpoint [8]dynamic: got message 74 to -1/-1 from 15/1
	Known components:
		241/190
Endpoint [8] accepted message 74 to -1/-1 from 15/1
TCP [8]dynamic: Wrote 29 bytes

…
…


```

### Test case 5: add GCS's ID
**OK, OUT filter works at the same time as OUT and IN filter. Let’s add 233 to allowed and see what happens.**

Configuration file content /etc/mavlink-router/main.conf :
```
[General]
TcpServerPort = 5760
[UartEndpoint ToRover15]
Device = /dev/ttyTHS1
Baud = 57600
AllowSrcSysOut = 15,233
```

**Test 5 result:** now QGroundControl SysID 233 gets full two-way communication with Rover 15, but QGroundControl SysID 241 only sees telemetry and unable to retrieve parameters (as in the Test 4). 
**My initial assumption is proved – AllowSrcSysOut acts as a two-way whitelist filter.**

Debug console output: sudo mavlink-routerd -g debug
```
mavlink-router version v2-229-g022333d
Opened UART [4]ToRover15: /dev/ttyTHS1
UART [4]ToRover15: speed = 57600
Opened TCP Server [5] [::]:5760
> UART [4]ToRover15: Got 40 bytes
Endpoint [4]ToRover15: got message 30 to -1/-1 from 15/1
	Known components:
		15/1
Message 30 to unknown sysid/compid: -1/-1
> UART [4]ToRover15: Got 40 bytes
Endpoint [4]ToRover15: got message 30 to -1/-1 from 15/1
	Known components:
		15/1
Message 30 to unknown sysid/compid: -1/-1
…
…
**> TCP [7]dynamic: Got 21 bytes**
Endpoint [4]ToRover15: got message 0 to -1/-1 from 233/190
	Known components:
		15/1
Endpoint [4] accepted message 0 to -1/-1 from 233/190
**UART [4]ToRover15: Wrote 21 bytes**
Endpoint [7]dynamic: got message 0 to -1/-1 from 233/190
	Known components:
		233/190
Endpoint [8]dynamic: got message 0 to -1/-1 from 233/190
	Known components:
		241/190
Endpoint [8] accepted message 0 to -1/-1 from 233/190
TCP [8]dynamic: Wrote 21 bytes
> UART [4]ToRover15: Got 82 bytes
Endpoint [4]ToRover15: got message 340 to -1/-1 from 15/1
	Known components:
		15/1
Endpoint [7]dynamic: got message 340 to -1/-1 from 15/1
	Known components:
		233/190
Endpoint [7] accepted message 340 to -1/-1 from 15/1
TCP [7]dynamic: Wrote 82 bytes
Endpoint [8]dynamic: got message 340 to -1/-1 from 15/1
	Known components:
		241/190
Endpoint [8] accepted message 340 to -1/-1 from 15/1
TCP [8]dynamic: Wrote 82 bytes

…
…
> TCP [8]dynamic: Got 44 bytes
Endpoint [4]ToRover15: got message 76 to 15/1 from 241/190
	Known components:
		15/1
**Endpoint [4] filtered out message 76 to 15/1 from 241/190**
Endpoint [7]dynamic: got message 76 to 15/1 from 241/190
	Known components:
		233/190
Endpoint [8]dynamic: got message 76 to 15/1 from 241/190
	Known components:
		241/190
> UART [4]ToRover15: Got 40 bytes
Endpoint [4]ToRover15: got message 33 to -1/-1 from 15/1
	Known components:
		15/1
Endpoint [7]dynamic: got message 33 to -1/-1 from 15/1
	Known components:
		233/190
Endpoint [7] accepted message 33 to -1/-1 from 15/1
TCP [7]dynamic: Wrote 40 bytes
Endpoint [8]dynamic: got message 33 to -1/-1 from 15/1
	Known components:
		241/190
Endpoint [8] accepted message 33 to -1/-1 from 15/1
TCP [8]dynamic: Wrote 40 bytes
…
…


```

### Test case 6: add all SysIDs
**OK, lets add all SysID to allowed and see what happens.**

Configuration file content /etc/mavlink-router/main.conf :
```
[General]
TcpServerPort = 5760
[UartEndpoint ToRover15]
Device = /dev/ttyTHS1
Baud = 57600
AllowSrcSysOut = 15,233,241
```

**Test 6 result: Everything works fine.** Can easily connect from both instances of QGroundControl (SysID 233 and 241 at the same time). Parameters download procedure is very fast. Can upload and download missions.

**My initial assumption is once again proved – AllowSrcSysOut acts as a two-way whitelist filter.**

**P.S.** Also tried various scenarios using several rovers and intermediate mavlink-routers (x86, arm), various combinations of endpoints TCP clients, UDP servers and clients, UART ports – all points out that AllowSrcSysIn filter does not work, but AllowSrcSysOut works as a two-way whitelist filter.

Thank you

Regards

Max

**Comment by jbeyerstedt** [comment]
Thanks for the detailed analysis. It was a simple copy-paste error when passing the config to the endpoint configuration. Sorry for that.

**Comment by max-itsl** [comment]
Good day! Thank you! Will test today.

Regards
Max
